### PR TITLE
A bunch of wrapping up things

### PR DESCRIPTION
**Pull request by Mr4k**

### CODE REVIEW

**Comment by Mr4k** [comment]
This is not a great pr and probably needs more testing but would be good to merge in I think

File: ast/ast.go [hunk 1]
```diff
@@ -6,10 +6,39 @@ import (
...
+	empty     bool
+	BeginLine int
+	BeginPos  int
+}
+
+func spanFromToken(t token.Token) Span {
+	return Span{BeginLine: t.Context.LineNumber, BeginPos: t.Context.PositionInLine}
+}
+
+func (self Span) merge(other Span) Span {
```
Comment (jemmaissroff) [inline ast/ast.go]: I think idiomatic go would be `s` or `span` instead of `self`

File: ast/ast.go [hunk 2]
```diff
@@ -24,6 +53,53 @@ type Expression interface {
...
+
+func (l *LengthNode) String() string {
+	return "len(" + l.Child.String() + ")"
+}
+
+func (l *LengthNode) Span() Span {
+	return l.Child.Span()
+}
+
+type OffsetNode struct {
```
Comment (jemmaissroff) [inline ast/ast.go]: Can we remove `OffsetNode`?

File: evaluator/builtins.go [hunk 3]
```diff
@@ -93,6 +94,23 @@ func init() {
 				return &res
 			},
 		},
+		"dep_diagraph": {
+			// NOTE this function is legacy to support tests from the old version
```
Comment (jemmaissroff) [inline evaluator/builtins.go:98]: Can we remove this comment?

File: evaluator/builtins.go [hunk 4]
```diff
@@ -93,6 +94,23 @@ func init() {
 				return &res
 			},
 		},
+		"dep_diagraph": {
```
Comment (jemmaissroff) [inline evaluator/builtins.go:97]: Could we name this `dependency_graph`?

File: evaluator/evaluator.go [hunk 5]
```diff
@@ -355,6 +399,8 @@ func addElements(left *object.Array, right *object.Array) *object.Array {
 	// NOTE (Peter) this should be okay instead of calling object.CreateArray
 	// But be very careful when changing this for dependency reasons
 	res := object.Array{}
+	res.Offset.ASTCreator = &ast.StringLiteral{Value: "OFFSET"}
```
Comment (jemmaissroff) [inline evaluator/evaluator.go]: Do we need this?

File: evaluator/evaluator_test.go [hunk 6]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 func testEval(input string) object.Object {
-	l := lexer.New(input)
+	l := lexer.New(input, "test_file.koko")
```
Comment (jemmaissroff) [inline evaluator/evaluator_test.go:11]: Thanks for catching all these ❤️

File: object/graphviz_utils.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		}
+	}
+	return out
+}
+
+func getObservableDepsFromObj(obj Object) []Object {
+	out := []Object{}
+	for d := range obj.GetDependencyLinks() {
+		if off, ok := d.(*Offset); ok {
+			for dep := range collapseOffsetNodes(off) {
```
Comment (jemmaissroff) [inline object/graphviz_utils.go:42]: Can this be: `out = append(out, collapseOffsetNodes(off)...)`?

File: object/graphviz_utils.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+func GetAllDependenciesToDotLang(result Object) string {
+	seenNodes := make(map[Object]bool)
+	queue := []Object{}
+	queue = append(queue, result)
+	out := "digraph G {\n"
+	seenOutputEdges := make(map[string]bool)
+	for len(queue) > 0 {
+		head := queue[0]
+		fmt.Printf("HEAD: %+v\n", head)
```
Comment (jemmaissroff) [inline object/graphviz_utils.go]: Can we remove the print?

File: object/object.go [hunk 9]
```diff
@@ -62,9 +62,14 @@ func GetAllDependencies(result Object) map[Object]bool {
 		}
 		out[head] = true
 		if len(queue) > 1 {
-			queue = append(queue[1:], head.GetDependencyLinks()...)
+			queue = queue[1:]
+			for link := range head.GetDependencyLinks() {
```
Comment (jemmaissroff) [inline object/object.go:66]: `queue = append(queue, head.GetDependencyLinks()...)`?

File: object/object.go [hunk 10]
```diff
@@ -62,9 +62,14 @@ func GetAllDependencies(result Object) map[Object]bool {
...
+				queue = append(queue, link)
+			}
 		} else {
-			queue = head.GetDependencyLinks()
+			for link := range head.GetDependencyLinks() {
```
Comment (jemmaissroff) [inline object/object.go:70]: Same as above

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: ast/ast.go (hunk 1 above)
Comment (Mr4k) [reply ast/ast.go]: 👍

File: evaluator/builtins.go (hunk 3 above)
Comment (Mr4k) [reply evaluator/builtins.go:98]: 👍

File: evaluator/evaluator.go (hunk 5 above)
Comment (Mr4k) [reply evaluator/evaluator.go]: noooo we don't

File: evaluator/evaluator_test.go (hunk 6 above)
Comment (Mr4k) [reply evaluator/evaluator_test.go:11]: 💯

File: object/graphviz_utils.go (hunk 7 above)
Comment (Mr4k) [reply object/graphviz_utils.go:42]: actually was wrong about this one. It can't b/c it's a dictionary

File: object/graphviz_utils.go (hunk 8 above)
Comment (Mr4k) [reply object/graphviz_utils.go]: for sure

File: object/object.go (hunk 9 above)
Comment (Mr4k) [reply object/object.go:66]: same problem re dictionary

File: object/object.go (hunk 10 above)
Comment (Mr4k) [reply object/object.go:70]: same problem re dictionary

File: evaluator/builtins.go (hunk 4 above)
Comment (Mr4k) [reply evaluator/builtins.go:97]: I'm gonna hold off here b/c it breaks my blog post (haha we already have tech debt)

**Pull request merged**